RBS additions

larskanis opened the pull request:
Happy new year as well @ParadoxV5 !

I did some minor adjustments to the CI to make it green. Then I checked the RBS files against the specs and was surprised about the many failures. So I went further to fix them in the subsequent commits.

I used the following command line:
```sh
RBS_TEST_TARGET='FFI::*' RUBYOPT='-rrbs/test/setup'  bundle exec rake spec
```

I was able to fix most RBS issues, but a few specs fail with mismatch in `Struct#layout` . I wasn't able to fix it properly. And `Function#attach`  fails with endless recursion. Probably an incompatibility with `super` call and the [`prepend` method](https://github.com/ffi/ffi/blob/f2d384690f9e27423beb34e685970c6a18b9acef/lib/ffi/function.rb#L69)

Could you please check my additions?

larskanis committed:
Exclude rbs.gem install from rubies that are not supported

larskanis committed:
CI: Print mkmf.log on failure on Appveyor

larskanis committed:
Avoid rb_class_new_instance() here, to avoid passing the method block

rb_class_new_instance() automatically passes the block forward to the initialize method.
That way the block of a callback is forwarded to the Struct initializer, like here in the specs:

      s = LibTest::S8F32S32.new
      ret = LibTest.testCallbackVrT { s }

Struct#initialize ignores the block, so that the code works, but it doesn't make sense.

This stood up when enabling rbs type checks, which complained about an unspecified block to Struct#initialize.

larskanis committed:
Fix documentation about Pointer

larskanis committed:
Various adjustments to rbs definitions

They stood up when running with rbs type checks enabled.

larskanis committed:
Remove double definition of Pointer#to_ptr

ParadoxV5 commented on sig/ffi/enum.rbs:
```diff
@@ -1,22 +1,22 @@
   class Enums
-    def initialize: ...
-    
+    def initialize: () -> instance
```
```suggestion
    def initialize: () -> void
```

`.new` returns `instance`, `#initialize`’s return doesn’t matter.

When I did my batch, I looked at sources briefly and typed some returns; e.g., all C-implemented `#initialize` returns `self`.
If we prefer to hide these minute details, we should use `-> void`.

Also applies to other classes’ `#initialize`.

ParadoxV5 commented on sig/ffi/enum.rbs:
```diff
@@ -1,22 +1,22 @@
-    
+
     def []: (Symbol query) -> Integer
-          | (Integer query) -> Symbol
+          | (Integer query) -> Symbol?
```
If so, may as well?
```suggestion
    def []: (Symbol query) -> Integer?
          | (Integer query) -> Symbol?
```

RBS for Ruby Core classes like `::Array` does it like this:
```suggestion
    def []: %a{implicitly-returns-nil} (Symbol query) -> Integer
          | %a{implicitly-returns-nil} (Integer query) -> Symbol
```

larskanis replied:
Can you please decide on your own? I only read a crash course about RBS and static typing...

ParadoxV5 replied:
`%a{implicitly-returns-nil}` doesn’t work with RBS Tests.

ParadoxV5 commented on sig/ffi/function.rbs:
```diff
@@ -11,13 +12,14 @@ module FFI
+        ?Hash[Symbol, untyped] options
       ) -> self
 
+#        ?blocking: boolish?, ?convention: Library::convention?, ?enums: Enums?
```
See ruby/rbs#1639

For now, it’s more intuitive if there isn’t a blank line gapping `#initialize` and this comment.

ParadoxV5 commented on sig/ffi/pointer.rbs:
```diff
@@ -28,7 +27,6 @@ module FFI
     def read_string_length: (Integer len) -> String
     def read_string_to_null: () -> String
     def slice: (Integer offset, Integer length) -> Pointer
-    def to_ptr: () -> self
```
This is to indicate that `Pointer#to_ptr` returns `self` (which *is* a `Pointer`).

ParadoxV5 commented on sig/ffi/pointer.rbs:
```diff
@@ -9,10 +9,9 @@ module FFI
+    def initialize: (?AbstractMemory::type_size type, (Pointer | Integer) address) -> instance
     def +: (Integer offset) -> Pointer
-    def ==: (Pointer other) -> bool
+    def ==: (Pointer? other) -> bool
```
Always good to have a second pair of eyes :+1:. I forgot that `nil` is accepted whereëver `Pointer`/`_ToPtr` are.

ParadoxV5 commented on sig/ffi/struct_by_reference.rbs:
```diff
@@ -3,7 +3,7 @@ module FFI
     attr_reader struct_class: Struct[P, untyped]
 
-    def initialize: (Struct[P, untyped] struct_class) -> void
+    def initialize: (S struct_class) -> instance
```
This change to `S` is correct. `#initialize` is called on the instance, so it has the type vars like `S` in scope.

ParadoxV5 commented on sig/ffi/abstract_memory.rbs:
```diff
@@ -159,6 +159,6 @@ module FFI
     def write_array_of_float: (Array[Numeric] ary) -> self
     def write_array_of_double: (Array[Numeric] ary) -> self
-    def write_array_of_pointer: (Array[Pointer::_ToPtr] ary) -> self
+    def write_array_of_pointer: (Array[Pointer::_ToPtr?] ary) -> self
```
Given that `nil` is also accepted whereëver `_ToPtr` is, would `type pointer = _ToPtr?` in `Pointer` RBS be a valuable addition?

RBS builtins has things like `type int = Integer | _ToInt`, albeit they’re redundant.

larskanis replied:
Can you please decide on your own?

ParadoxV5 replied:
It’s implicit, but `#(put|write)_array_of_pointer` also accepts Integer elements (*but not `#to_int`s* since they share the conversion implementation with `#(put|write)_pointer`.

---

While [FFI currently only mention `#to_ptr` in `AbstractMemory`](https://github.com/search?q=repo%3Affi%2Fffi%20%22%23to_ptr%22&type=code), the `_ToPtr?|Integer` pattern is useful in user libraries’ (i.e., bindings based on FFI) RBS as well, so I’ll surface the type alias to the `FFI` namespace.

ParadoxV5 commented on sig/ffi/abstract_memory.rbs:
```diff
@@ -107,7 +107,7 @@ module FFI
     def get_array_of_float32: (Integer offset, Integer length) -> Array[Float]
     def get_array_of_float64: (Integer offset, Integer length) -> Array[Float]
     def get_array_of_pointer: (Integer offset, Integer length) -> Array[Pointer]
-    def get_array_of_string: (Integer offset, Integer length) -> Array[String]
```
Hah, the copypaste work got me.

How about (in the future) we let a Rake task or two generate `abstract_memory.rbs` and `native_type.rbs`?

larskanis replied:
It's OK for me without a rake task, but I'm not opposed to it.

ParadoxV5 commented on sig/ffi/function.rbs:
```diff
@@ -1,6 +1,7 @@
 module FFI
   interface _Function
-    def attach: (Module mod, String name) -> self
+    # TODO: leads to a endless recursion when used with -rrbs/test/setup
+    # def attach: (Module mod, String name) -> self
```
Perhaps porting `Function#attach` to Ruby (similar to its cousin, `VariadicInvoker#attach`) and flatten the `RegisterAttach` private module away would solve/workaround this problem.

> Did you know that even CRuby is migrating some implementations from C to Ruby?
> With all the JIT improvements, Ruby code’s performance is rivaling that of C itself!

In fact, it makes sense to me if `Function` and `VariadicInvoker` share some implementations with a common ancestor module, or even (if past design decisions aren’t in the way) have `VariadicInvoker` inherit from `Function`.

larskanis replied:
I opened an issue regarding the [stack level too deep](https://github.com/ruby/rbs/issues/1710).

ParadoxV5 reviewed (commented):
By `Struct#layout` you meant `Struct.layout` (AKA `Struct::layout`).

Test failures show that a `[Symbol, Integer]` 2-tuple is also valid to come after a member name. This wasn’t [documented](https://rubydoc.info/gems/ffi/1.16.3/FFI%2FStruct.layout).
> ```ruby
> layout :a, :int, :c, :test_enum,
>   :d, [ TestEnum, TestEnum.symbols.length ]
> ```
> ```ruby
> layout( :a, [:char, 10],
>         :i, :int,
>         :f, :float,
>         :d, :double,
>         :s, :short,
>         :l, :long,
>         :j, :long_long,
>         :c, :char )
> ```

It’d be more consistent (to type check with, as well) if the array form has each entry be a 2–3 tuple in the form `[Symbol name, layout_type type, ?Integer offset]`.

[Update]
I thought the integer in the 2-tuple is the offset, but the tuple actually stands for an array, with the first element as the field type and the interger the array size.
This is [documented in “Char arrays” in the wiki](https://github.com/ffi/ffi/wiki/Structs/234eab91d0ee55ce103d4b92a96a98c6e8e46890#char-arrays), though the array can be of any type.

ParadoxV5 committed:
Fix `Struct::layout`

ParadoxV5 commented on sig/ffi/library.rbs:
```diff
@@ -25,15 +25,15 @@ module FFI
     def freeze: () -> void
     def function_names: (_ToS name, Array[Type | singleton(Struct)] arg_types) -> Array[String]
-    def typedef: [T < Type] (T old, Symbol add, ?untyped) -> T
+    def typedef: [T < Type] (T old, (Symbol | DataConverter) add, ?untyped) -> T
```
```suggestion
    def typedef: [T < Type] (T old, Symbol | DataConverter add, ?untyped) -> T
```

larskanis committed:
Move typedef, type_find and related methods around

For better readability.
And add some comments.

larskanis committed:
Avoid rb_class_new_instance() here, to avoid passing the method block

Similar to commit 5b4b4060727f7ade3842011350635370970ef650

rb_class_new_instance() automatically passes the block forward to the initialize method.
That way the block of an array function is forwarded to the Struct initializer.
Struct#initialize ignores the block, so that the code works, but it doesn't make sense.

This stood up when enabling rbs type checks, which complained about an unspecified block to Struct#initialize.

larskanis committed:
RBS: Type maps can also have DataConverters as key

larskanis committed:
RBS: custom_typemap is private

larskanis committed:
RBS: Fix errors in puts kind of:

  11) MemoryPointer allows writing as a sized int
      Failure/Error: put(type, 0, value)

      RBS::Test::Tester::TypeError:
        TypeError: [FFI::AbstractMemory#put] ArgumentTypeError: expected `bot` (value) but given `1`

larskanis committed:
RBS: Fix errors in write kind of:

  11) MemoryPointer allows writing as a sized int
      Failure/Error: m.write :uint32, 1

      RBS::Test::Tester::TypeError:
        TypeError: [FFI::Pointer#write] ArgumentTypeError: expected `bot` (value) but given `1`
      # /home/lars/.rvm/gems/ruby-3.3.0/gems/rbs-3.4.1/lib/rbs/test/tester.rb:158:in `call'
      # /home/lars/.rvm/gems/ruby-3.3.0/gems/rbs-3.4.1/lib/rbs/test/observer.rb:10:in `notify'
      # /home/lars/.rvm/gems/ruby-3.3.0/gems/rbs-3.4.1/lib/rbs/test/hook.rb:178:in `ensure in write__with__RBS_TEST_eb5149_4888829b'
      # /home/lars/.rvm/gems/ruby-3.3.0/gems/rbs-3.4.1/lib/rbs/test/hook.rb:178:in `write__with__RBS_TEST_eb5149_4888829b'
      # ./spec/ffi/rbx/memory_pointer_spec.rb:65:in `block (2 levels) in <top (required)>'
      # ------------------
      # --- Caused by: ---
      # RBS::Test::Tester::TypeError:
      #   TypeError: [FFI::AbstractMemory#put] ArgumentTypeError: expected `bot` (value) but given `1`
      #   /home/lars/.rvm/gems/ruby-3.3.0/gems/rbs-3.4.1/lib/rbs/test/tester.rb:158:in `call'

larskanis committed:
RBS: Integer values are converted by to_int

larskanis committed:
RBS: Struct#order(Symbol) returns a Struct whereas order() returns Symbol

larskanis committed:
RBS: declare initialize methods as returning void

By convention as discussed in https://github.com/ParadoxV5/ruby-ffi/pull/1#discussion_r1440050785

larskanis committed:
RBS: remove parantheses as suggested

https://github.com/ParadoxV5/ruby-ffi/pull/1/files#r1440105066
